### ISSUE TITLE
Crash when reading pcap files with an ARP packet.

### ISSUE THREAD

**Issue by 0xmitsurugi** [the-tcpdump-group/tcpdump#423]
I manage to get some crashes when reading pcap file containing one ARP packet in 
hex_and_ascii_print_with_offset() function.

Info:
```bash
mitsurugi@mitsu:~/fuzz/tcpdump-4.6.2$ ./tcpdump --version
tcpdump version 4.6.2
libpcap version 1.3.0
OpenSSL 1.0.1e 11 Feb 2013
mitsurugi@mitsu:~/fuzz/tcpdump-4.6.2$
mitsurugi@mitsu:~/fuzz/tcpdump-4.6.2$ cat cra.sh
#! /bin/bash

#global header
  #Magic number -- Mandatory
  printf '\xd4\xc3\xb2\xa1'
  #Version major + version minor (4B+4B) -- Can be anything
  printf '\x04\x00' ; printf '\x02\x00'
  #related to time : epoch + microsecs -- Can be anything
  printf '\x4d\x69\x74\x73\x75\x72\x75\x67'
  #max length of packets -- usually 65535
  printf '\x69\xff\x00\x00'
  #link layer type  -- important
  printf '\x01\x00\x00\x00'

#header Packet 1
  #epoch time+microsecs -- Can be anything
  printf '\x74\x68\x65\x20\x6e\x61\x6d\x65'
  #size of saved packet
  printf '\x16\x00\x00\x00' # 0x16 == 22 -- Must comply with saved pkt data
     #and must be > 0x15 to crash
  #Size of packet data on the wire
  printf '\x9f\x06\x02\x00'  #if too little: no crash
     # Can't figure why this number

#data packet 1
  # two mac addresses: -- can be random
  printf '\x00\x00\xb3\x00\x00\x30\x00\x00\x00\x00\x00\x01'
  # ethertype ARP: -- Mandatory to crash
  printf '\x08\x06' 
  # Hardware type+Proto type: -- does not seem important
  printf '\x00\x01'; printf '\x08\x00'
  # HW addr len 1B + Proto addr len 1B + Operation 2B
  # HW addr len must be >1 to crash
  printf '\x02\x00\x00\x00'
  # data -- just fill it to comply with size of save pkt
  printf '\x00'
mitsurugi@mitsu:~/fuzz/tcpdump-4.6.2$ ./cra.sh > file.pcap
mitsurugi@mitsu:~/fuzz/tcpdump-4.6.2$ ./tcpdump -r file.pcap > /dev/null
reading from file file.pcap, link-type EN10MB (Ethernet)
Segmentation fault
mitsurugi@mitsu:~/fuzz/tcpdump-4.6.2$
```

I used the redirection to /dev/null because the full printing is enormous (more than 8000 lines of output).
Here is the end of a gdb trace:
gdb output:
```gdb
Program received signal SIGSEGV, Segmentation fault.
--------------------------------------------------------------------------[regs]
  EAX: 0x0000002E  EBX: 0xBFFFDDB5  ECX: 0x0000000F  EDX: 0x0833C000  o d I t S z a p c 
  ESI: 0xBFFFDD8F  EDI: 0x00000001  EBP: 0x0833C002  ESP: 0xBFFFDD40  EIP: 0x08074097
  CS: 0073  DS: 007B  ES: 007B  FS: 0000  GS: 0033  SS: 007B
--------------------------------------------------------------------------[code]
=> 0x8074097 <hex_and_ascii_print_with_offset+623>:	movzx  eax,BYTE PTR [edx]
   0x807409a <hex_and_ascii_print_with_offset+626>:	sub    ecx,0x1
   0x807409d <hex_and_ascii_print_with_offset+629>:	mov    DWORD PTR [esp+0x38],ecx
   0x80740a1 <hex_and_ascii_print_with_offset+633>:	movzx  ecx,BYTE PTR [edx+0x1]
   0x80740a5 <hex_and_ascii_print_with_offset+637>:	mov    BYTE PTR [esp+0x34],al
   0x80740a9 <hex_and_ascii_print_with_offset+641>:	mov    eax,DWORD PTR [esp+0x2c]
   0x80740ad <hex_and_ascii_print_with_offset+645>:	mov    DWORD PTR [esp+0x8],0x820da40
   0x80740b5 <hex_and_ascii_print_with_offset+653>:	mov    DWORD PTR [esp],esi
--------------------------------------------------------------------------------
hex_and_ascii_print_with_offset (ndo=ndo@entry=0x8319880, ident=ident@entry=0x821a08c "\n\t", cp=<optimized out>, cp@entry=0x831b98e "", length=length@entry=0x20691, oset=0x20670, oset@entry=0x0) at ./print-ascii.c:102
102			s1 = *cp++;
gdb$
```

This has been found with afl, American Fuzzy Lop

**Comment by 0xmitsurugi** [comment]
Sorry for the noise: This one seems corrected when I download the last tcpdump today from https://github.com/the-tcpdump-group/tcpdump/archive/master.zip 

```bash
$ ./tcpdump --version
tcpdump version 4.7.0-PRE-GIT_2015_02_02
libpcap version 1.3.0
OpenSSL 1.0.1e 11 Feb 2013
$ ./tcpdump -r file.pcap 
reading from file file.pcap, link-type EN10MB (Ethernet)
17:53:08.170166718 ARP, Ethernet (len 2), IPv4 (len 0), length 132753
tcpdump: pcap_loop: truncated dump file; tried to read 16 header bytes, only got 1
$ 
```

I think you can close this one.

**Comment by infrastation** [comment]
This may be a duplicate of #417.

**Comment by mcr** [comment]
Do we have a pcap test file for this arp? 0xmitsurugi, would you like to change this into a pull request with additional test code?

**Comment by 0xmitsurugi** [comment]
The cra.sh shell script given creates the pcap file.
Just copy paste it, and:
$ ./cra.sh > file.pcap
does the job.

**Comment by infrastation** [comment]
For history: this issue was fixed by commit 4ac7226 on 23 Oct 2014.
